### ISSUE TITLE
[DDP] Process that runs pytorch lightning trainer SIGKILLs its peers

### ISSUE THREAD

**Issue by aivanou** [PyTorchLightning/pytorch-lightning#9388]
## 🐛 Bug

When exception occurs, logic in [trainer](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L525) calls `reconciliate_processes`, which resolves for DDP plugin in performing [SIGKILL](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L505) on all peer processes. 

This functionality does not work well if trainers are executed under supervision of a parent process, e.g. `[torchelastic](https://pytorch.org/docs/stable/distributed.elastic.html)` library, or any scheduler framework. 

With this functionality, it is hard for the parent process that controls trainers to properly propagate an error, since there is a race condition occurs:

The job may fail with only peer process that got killed via SIGKILL, or the job may fail recording both the peer process and the process that sent the SIGKILL. 

I was wondering whether it is possible to revise this piece of functionality, e.g. in ddp3, where peers do not perform SIGKILL? 
For example, to fail the process, and rely on the parent process to actually perform termination, or use already created [store](https://pytorch.org/docs/stable/distributed.html#distributed-key-value-store) to notify peer processes to shutdown? 

My apologies, if this is supposed to be in a `feature` section. I was not sure whether it is a bug or a feature. Also, let me know if I need to create [RFC] with proposed solution for the issue.

**Comment by ananthsub** [comment]
@awaelchli @tchaton  - one thing we could do is only enable `reconciliate_processes` if the cluster environment doesn't create processes. this way torchelastic/slurm would still rely on the parent process for termination: https://github.com/PyTorchLightning/pytorch-lightning/blob/41ba639859cf6c6bf319eb33e5b3394504315962/pytorch_lightning/plugins/environments/torchelastic_environment.py#L34-L35

**Comment by tchaton** [comment]
Dear @aivanou,

Thanks raising this issue. 
When I designed this functionally, I thought about this use-case and I believe it is working as I intented to work.

The `reconciliate_processes` would kill the others only when they haven't reconciliated (meaning other processes should be waiting on a barrier which would never been passed).

This mechanism is there to ensure your staled scripts gets kill instead of waiting for ever. This was a long requested by several users to prevent extra costs while running training in the cloud.

When using torchelastic, in my experiment, it seems to work correctly. The killed processed will be marked as Terminated and the killing process will raise a `DeadlockDetectedException`. I didn't see any lost logs from TorchElastic side.

Another approach I investigated was to set a timeout to the TCPStore, unfortunately this would trigger a RuntimeError and we won't be able to gracefully exit and save a fault tolerant checkpoint.

Note: Something helpful would be for Torchelastic to expose a functionality to kill processes.

Best,
T.C

**Comment by aivanou** [comment]
Hi Thomas (@tchaton )

Thank you very much for the detailed response!

Right now Torchelastic agent does not terminate processes across the cluster, but only processes that it manages on a single host.

>>> This was a long requested by several users to prevent extra costs while running training in the cloud.
It would be great if can point me to any conversations around that, I am interested in how they are running training in the cloud. 

The logic of SIGKILL the peer process works in terms of terminating the job: the training job will be terminated. But this logic can hide the real exception that occurred during the job execution:

E.g. Process A sends SIGKILL to process B. Process B fails, and scheduler that controls the job, just terminates the whole job without waiting for process A to terminate. The whole job will be terminated due to `Process B got SIGKILL`, while the real reason will be never found (`Process A got an error`). This makes distributed jobs very hard to debug. 

We can enable torchelastic to terminate jobs across the cluster, but I am sure not what other supported cluster environments will do. 

I think if at least one trainer process gets an exception, another trainer processes should not wait on the barrier and just terminate. But I think it would require introduction of some kind of heartbeat functionality, e.g. background thread that pings every N seconds the same `Store` as used to perform barrier. It would be great to know if you remember any examples that I can look in.

In general, I think if cluster environment is set, it is up to the parent framework that started the process to terminate trainers.

**Comment by tchaton** [comment]
Hey @aivanou,

You are entirely right, an heartbeat mechanism would be a great approach.

In that case, the best solution for the short term would be to prevent killing the process when the cluster creates the processes.

Best,
T.C

**Comment by awaelchli** [comment]
@ananthsub thanks for helping with this issue
